An SMT2 frontend plus solver binary

kroening opened the pull request.

kroening committed:
fix line number counting in SMT2 tokenizer

kroening committed:
a translator from SMT2 into expressions

kroening committed:
an SMT2 solver using boolbv and satcheckt

kroening commented:
Note that the actual solver doesn't link because of a pending dependency cleanup.

Will add some tests for the solver binary; suggestions for a sensible choice (not all of SMT-LIB) are welcome.

tautschnig commented on src/solvers/smt2/smt2_solver.cpp:
```diff
@@ -0,0 +1,91 @@
+#include <fstream>
```
No copyright information.

kroening replied:
Done

tautschnig commented on src/solvers/Makefile:
```diff
@@ -215,7 +215,7 @@ INCLUDES += -I .. \
 CLEANFILES = solvers$(LIBEXT)
 
-all: solvers$(LIBEXT)
+all: solvers$(LIBEXT) smt2_solver$(EXEEXT)
```
The clean target, .gitignore, and CMake will need updates.

tautschnig reviewed (changes requested):
1. Placing this new front-end in solvers/ requires a fair bit of further work (see also comments), and the way it is set up it will always be built when anyone depends on solvers, which seems rather unnecessary.
2. When refactoring, please add comments (at least at the level of classes).

kroening commented:
yes, I am contemplating to not make the smt2_solver binary a default target; however, this would be the first such instance.

kroening committed:
tests for smt2_solver

kroening committed:
added to_mathematical_function_type

kroening committed:
add support for let_exprt

kroening commented:
The cleanup is taking too long, will aim at getting this merged before.

tautschnig commented:
> The cleanup is taking too long, will aim at getting this merged before.

The cleanup of what? And which PR should be merged before which?

kroening commented:
A cleanup of the dependencies inside solver/; removing these enables the build of the smt2_solver binary without linking to goto-programs, langapi, and java_bytecode.

polgreen commented:
There is something wrong with the let parser. The smt benchmarks in this folder fail: 
https://clc-gitlab.cs.uiowa.edu:2443/SMT-LIB-benchmarks/QF_BV/tree/master/2017-BuchwaldFried

tautschnig commented on .gitignore:
```diff
@@ -104,6 +104,7 @@ src/goto-instrument/goto-instrument.exe
 src/jbmc/jbmc
 src/musketeer/musketeer
 src/musketeer/musketeer.exe
+src/solvers/smt2/smt2_solver
```
Even though a later commit disables this: the .exe variant should be added now.

kroening replied:
done

tautschnig commented on regression/Makefile:
```diff
@@ -16,6 +16,7 @@ DIRS = ansi-c \
        invariants \
        strings \
        jbmc-strings \
+#       smt2_solver \
```
What's the effect of having a commented-out line ending in a `\` in a block of lines that end in `\`? (It certainly confuses GitHub's syntax highlighting, but possibly also `make`.)

kroening replied:
Removed

tautschnig commented on src/solvers/flattening/boolbv.cpp:
```diff
@@ -463,6 +459,15 @@ literalt boolbvt::convert_rest(const exprt &expr)
+    bvt bv=convert_let(to_let_expr(expr));
+
+    if(bv.size()!=1)
+      throw "convert_let returned non-bool bitvector";
```
This should be an `INVARIANT`. Or, really, why does it not return a `literalt` instead?!

kroening replied:
There are both cases, i.e., one where let is used in a boolean context and one where it produces bit-vectors. Clarified the error message.

tautschnig commented on src/solvers/flattening/boolbv_let.cpp:
```diff
@@ -0,0 +1,28 @@
+
+  map.set_literals(id, expr.symbol().type(), value_bv);
+  bvt result_bv=convert_bv(expr.where());
+  map.erase_literals(id, expr.symbol().type());
```
Why are the first created/set and then erased?!

kroening replied:
Added a comment. The symbol is meant to be visible during the recursive call only.

tautschnig commented on src/solvers/flattening/boolbv_let.cpp:
```diff
@@ -0,0 +1,28 @@
+{
+  const bvt value_bv=convert_bv(expr.value());
+
+  // We expect the bindings to be unique, and
```
Bindings between ... what?

kroening replied:
Re-worded comment

tautschnig commented on src/solvers/flattening/boolbv_map.cpp:
```diff
@@ -146,3 +146,13 @@ void boolbv_mapt::set_literals(
+  map_entryt &map_entry=get_map_entry(identifier, type);
+
+  for(auto &l : map_entry.literal_map)
+    l.is_set=false;
```
This does not seem to "erase" literals?

kroening replied:
ok, it does now

tautschnig commented on src/solvers/prop/prop_conv.cpp:
```diff
@@ -327,8 +327,23 @@ literalt prop_conv_solvert::convert_bool(const exprt &expr)
+    {
+      literalt value=convert(let_expr.value());
+
+      // We expect the bindings to be unique, and
```
Bindings between what?

kroening replied:
Same as above, reworded

tautschnig commented on src/solvers/Makefile:
```diff
@@ -217,7 +217,7 @@ INCLUDES += -I .. \
 CLEANFILES = solvers$(LIBEXT) smt2_solver$(EXEEXT)
 
-all: solvers$(LIBEXT) smt2_solver$(EXEEXT)
+all: solvers$(LIBEXT) # smt2_solver$(EXEEXT)
```
The commit message does not say "why" it was disabled, and in particular this does not explain why earlier commits first enabled/added it while now it is being disabled.

kroening replied:
reworded

kroening commented:
The 2017-BuchwaldFried benchmarks now parse (they need back-end support for rol, but that's orthogonal).

kroening commented:
The Travis failure is spurious, of course.

peterschrammel committed:
Add missing langapi dependency to util

This dependency was introduced by moving language.h to langapi.
The dependency will be removed by the work in 1869 / 1935.

peterschrammel commented:
@kroening, cmake builds pass now.

tautschnig reviewed (approved):
I am not happy that a command-line tool now sits in the solvers/ directory, and is built every time, but this is a minor concern. Things look ok otherwise.

chrisr-diffblue reviewed (approved):
Not entirely sure why the solvers CMakeLists.txt needed to have CXX_STANDARD set rather than inheriting from the top-level CMakeLists properties, but it does appear it's needed. Other than that, I'm happy with the parts the parts for which I'm a code owner.
